### DISCUSSION TITLE
Welcome to ETSAuto Discussions!

### DISCUSSION THREAD

**Discussion by Lyric0620** [Lyric0620/ETSAuto#57] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by eiyooooo** [comment]
建议最高时速可以由用户通过键盘控制

**Comment by Lyric0620** [comment]
@eiyooooo 现在车道线保持的最高时速只能到75km/s，大于75km/s后卡车会开始摆动

**Reply by eiyooooo** [reply]
https://github.com/Yutong-gannis/ETSAuto/blob/ce4a9a649e55a51484741ea0bd0caff265d8af37/Condition/truck_condition.py#L25
这个改大一点可以减少摆动

还有这个没改过来
https://github.com/Yutong-gannis/ETSAuto/blob/ce4a9a649e55a51484741ea0bd0caff265d8af37/User/user_launcher.py#L49

**Comment by Lyric0620** [comment]
> https://github.com/Yutong-gannis/ETSAuto/blob/ce4a9a649e55a51484741ea0bd0caff265d8af37/Condition/truck_condition.py#L25
> 
> 
> 这个改大一点可以减少摆动
> 还有这个没改过来
> 
> https://github.com/Yutong-gannis/ETSAuto/blob/ce4a9a649e55a51484741ea0bd0caff265d8af37/User/user_launcher.py#L49

好的，ld你改过吗，改成多少

**Reply by eiyooooo** [reply]
破案了，摆动是因为车速和一个参数不关联
ld不用改
Truck里面的update函数少了行`self.ld = 1.0 * self.speed + self.lf`
然后将lf改为3`self.lf = 3  # 预瞄距离补偿`
效果好多了
![image](https://github.com/Yutong-gannis/ETSAuto/assets/52876220/00bf4b2d-787d-445c-ad48-4fd22d4b83d5)

**Reply by eiyooooo** [reply]
lf需要再调一调，3~4之间

**Reply by eiyooooo** [reply]
> lf需要再调一调，3~4之间

lf调3.63不错

**Comment by Shzzzz** [comment]
大佬，麻烦问下目标检测模块和车道线检测模块是怎么训练的呢，训练后的结果如何？

**Reply by Lyric0620** [reply]
目标检测采用的yolov8n模型，代码是官方仓库的，数据集采用的bdd100k，目前只训练了vehicle、truck、bus、traffic_light这些类别，输入图片尺寸是480x480的，训练配置基本没改，训练了100个epoch。

车道线检测用的bev-lanedet，我fork了这个仓库，可以在我主页找到。backbone用的resnet18，数据集用的openlane v1，输入大小大概是270x360，在dataloader中做了一些修改，即只关心左右两根车道，训练了30个epoch。

由于没有在ets上的数据集，所以两个模型均没有做测试。但两个模型的最终效果都在我可以容忍的范围之内，因为我只对前方60米、左右各4米的范围感兴趣，单帧推理加起来不会超过50ms。如果你想了解模型测试效果，可以用我发布的最新onnx模型进行测试，在项目中相应检测模块的文件夹下均有对视频的onnxruntime推理demo，可以进行参考。

**Reply by Shzzzz** [reply]
好的，谢谢大佬的回答，我自己去测试一下整个模型的效果。

**Reply by eiyooooo** [reply]
你好，请问一下之前的yolov5模型是怎么训练的？

**Reply by Lyric0620** [reply]
@eiyooooo 我训练的yolov8，方法应该是一样的。用的bdd100k数据集，先转换成yolo格式，写好yaml配置文件就可以开始训练了

**Reply by eiyooooo** [reply]
我是说之前v1那个版本的yolov5🤪

**Reply by Lyric0620** [reply]
@eiyooooo V1用的yolov6，也是一样的步骤。yolo几个版本都是一样的，转换数据集然后写配置文件

**Reply by eiyooooo** [reply]
好的，谢谢

**Comment by Shzzzz** [comment]
<img width="1133" alt="Snipaste_2023-10-07_21-26-42" src="https://github.com/Yutong-gannis/ETSAuto/assets/111844969/189317ae-13c8-43a3-a7bd-e8861c94c70e">
大佬麻烦问下，我这个按照read.me弄完，打开ETSAuto.bat文件，怎么是这样呢？

**Reply by Lyric0620** [reply]
左边是打开的各个进程，一共有六个，记录了一些log信息，也可以根据bat文件内容打开相应的python文件。报错是因为没有把游戏窗口放出来，没有检测到东西。你首先检查六个进程是否都打开了，然后把游戏窗口放屏幕左上角应该就行了。

**Reply by Shzzzz** [reply]
<img width="1280" alt="Snipaste_2023-10-07_21-40-14" src="https://github.com/Yutong-gannis/ETSAuto/assets/111844969/a019b1d6-d5fb-4fc7-be25-7a3bcdb4dff2">
我将you游戏打开了还是这样。

**Reply by Shzzzz** [reply]
![image](https://github.com/Yutong-gannis/ETSAuto/assets/111844969/b2e336cf-2b58-4b3d-b7d2-1379681eb525)
然后还保了个错。

**Reply by Lyric0620** [reply]
这个是我用except try已经跳过的报错，不用管。现在程序可以介入控制了吗

**Reply by Shzzzz** [reply]
还是不能，而且我发现那个dll文件我放在x84里面进游戏不会提示有sdk，放在x64里面才会提示。不知道是不是和这个有关，程序没有介入控制。

**Reply by Lyric0620** [reply]
这样，你先只打开perception_launcher.py、 planning_launcher.py、 user_launcher.py，看看感知能不能顺利跑起来。

**Reply by Lyric0620** [reply]
游戏画面不能遮挡住

**Reply by Lyric0620** [reply]
关于sdk是否工作正常，你可以加速卡车，看UI上的速度显示是否会改变

**Reply by Shzzzz** [reply]
运行perception_launcher.py报错：
onnxruntime.capi.onnxruntime_pybind11_state.NoSuchFile: [ONNXRuntimeError] : 3 : NO_SUCHFILE : Load model from E:\SteamLibrary\steamapps\common\Euro Truck Simulator 2\ETSAuto-2.2.1\weights\bevlanedet\resnet18_0.5_v2\ep049.onnx failed:Load model E:\SteamLibrary\steamapps\common\Euro Truck Simulator 2\ETSAuto-2.2.1\weights\bevlanedet\resnet18_0.5_v2\ep049.onnx failed. File doesn't exist
![image](https://github.com/Yutong-gannis/ETSAuto/assets/111844969/3d13e322-c946-487c-9d6d-c32f8a43ec7b)

**Reply by Shzzzz** [reply]
我已经把两个权重文件放入了weights文件夹
![image](https://github.com/Yutong-gannis/ETSAuto/assets/111844969/747b0998-d1e2-4d12-9a44-808ed5518d30)

**Reply by Lyric0620** [reply]
你把相应文件这里改成这样
https://github.com/Yutong-gannis/ETSAuto/blob/2f100014caa1a04d2da5380c4e1c145246252497/Selfdrive/Perception/perception_launcher.py#L26-L27

**Comment by Shzzzz** [comment]
大佬，我准备用下你的1.1版本，但是里面的环境感知pt权重文件那个链接失效了，还有这个文件吗？

**Reply by Lyric0620** [reply]
改了地址还是不行吗？

**Reply by Lyric0620** [reply]
1.1版本需要用到tensorrt，环境会比较麻烦

**Reply by Shzzzz** [reply]
环境已经搭好了

**Reply by Shzzzz** [reply]
我点那个链接，是404

**Reply by Lyric0620** [reply]
我找找

**Reply by Shzzzz** [reply]
行，谢谢大佬

**Reply by Lyric0620** [reply]
我重新发了一遍，你再试试 https://github.com/Yutong-gannis/ETSAuto/releases/tag/v1.1dev

**Reply by Shzzzz** [reply]
好像还是没有weather_scene_mobilenet_v3_412.pt这个文件

**Reply by Shzzzz** [reply]
看到tools里面的onnx2trt.py文件了，生成环境感知的engine文件了。大佬你的readme好像写的还是将weather_scene_mobilenet_v3_412.pt移动到Self-driving-Truck-in-Euro-Truck-Simulator2\Perception\SceneClassifier\weights文件夹下。

**Reply by Lyric0620** [reply]
就是这个文件，重命名一下 https://github.com/Yutong-gannis/ETSAuto/releases/download/v1.1dev/weather_detector.onnx

**Reply by Shzzzz** [reply]
![image](https://github.com/Yutong-gannis/ETSAuto/assets/111844969/3ed78dd8-0d26-4f33-b359-6a081c71cf70)
把这个文件重命名后，放入指定文件夹，然后运行报错了

**Reply by Lyric0620** [reply]
这几行注释掉，这个功能还没有用到
https://github.com/Yutong-gannis/ETSAuto/blob/2883ed30b63d9a8d5b58ab2a7e0b0ffcf369df81/script/main.py#L48

https://github.com/Yutong-gannis/ETSAuto/blob/2883ed30b63d9a8d5b58ab2a7e0b0ffcf369df81/script/main.py#L51

https://github.com/Yutong-gannis/ETSAuto/blob/2883ed30b63d9a8d5b58ab2a7e0b0ffcf369df81/script/main.py#L105-L108

**Reply by Shzzzz** [reply]
![image](https://github.com/Yutong-gannis/ETSAuto/assets/111844969/236a1b8a-47d9-4d3a-81ef-52000ec8ad6b)
我现在能在游戏中运行了，但是不知道为什么车道线检测好像没起作用，它在自动驾驶的时候会向右偏直到撞墙。

**Reply by Lyric0620** [reply]
这个版本是通过寻小地图运行的，需要先按F5把小地图调到最大，然后慢慢调整游戏窗口使效果最好

**Reply by Shzzzz** [reply]
那我还是试试2.2.1的版本吧

**Reply by Lyric0620** [reply]
2.2.1发布的时候我没改这两行，需要你自己改一下
https://github.com/Yutong-gannis/ETSAuto/blob/2f100014caa1a04d2da5380c4e1c145246252497/Selfdrive/Perception/perception_launcher.py#L26-L27

**Reply by Shzzzz** [reply]
我好像运行起来了，但是麻烦问下怎么在2.2.1版本上出现这个detect的窗口呢？就像你发的b站那个一样
![image](https://github.com/Yutong-gannis/ETSAuto/assets/111844969/ee32b2ed-ce9d-4996-81db-c0867996df3c)

**Reply by Lyric0620** [reply]
不好意思，目前版本没有提供原图和bev显示，可视化是将在虚拟相机下进行的。

**Reply by Shzzzz** [reply]
麻烦可以给我讲下怎么实现这个功能吗？

**Reply by Lyric0620** [reply]
你是需要在原图上画还是画bev

**Reply by Shzzzz** [reply]
在原图上画就行

**Reply by Shzzzz** [reply]
如果两个都能显示就最好了

**Reply by Lyric0620** [reply]
bev在v2.2.0上是有一个函数的，但是只能显示车道线，并且由于没有维护，需要自己debug
https://github.com/Yutong-gannis/ETSAuto/blob/ec251a4e0979137d2440717fda735722452f8ea1/User/interface.py#L110-L175
关于调用请参考这个文件
https://github.com/Yutong-gannis/ETSAuto/blob/v2.2.0/User/user_launcher.py

**Reply by Lyric0620** [reply]
关于原图显示其实可以在UserInterface上改
https://github.com/Yutong-gannis/ETSAuto/blob/2f100014caa1a04d2da5380c4e1c145246252497/Selfdrive/User/lib/interface.py#L15-L132
调整相机的旋转
https://github.com/Yutong-gannis/ETSAuto/blob/2f100014caa1a04d2da5380c4e1c145246252497/Selfdrive/User/lib/interface.py#L24
调整相机位置
https://github.com/Yutong-gannis/ETSAuto/blob/2f100014caa1a04d2da5380c4e1c145246252497/Selfdrive/User/lib/interface.py#L28
调整原图大小
https://github.com/Yutong-gannis/ETSAuto/blob/2f100014caa1a04d2da5380c4e1c145246252497/Selfdrive/User/lib/interface.py#L25
用原图替代空白画布
https://github.com/Yutong-gannis/ETSAuto/blob/2f100014caa1a04d2da5380c4e1c145246252497/Selfdrive/User/lib/interface.py#L67

原图需要从感知模块上保存到本地
参考这个类的publish
https://github.com/Yutong-gannis/ETSAuto/blob/37bb172e263b313daf5953eead4694566956d292/Screen/grab_screen.py#L9-L40
在update后调用publish
再在UserInterface里读取就行

**Reply by Lyric0620** [reply]
我不太建议你这样做，这样每秒要保存读取图片至少20次，会消耗掉很多计算资源

**Reply by Shzzzz** [reply]
了解了，谢谢大佬

**Reply by Shzzzz** [reply]
麻烦问下，您当时放在b站视频上跑的那个版本方便发我一份吗。我有点想学习一些怎么实现那个窗口检测的功能。

**Reply by Lyric0620** [reply]
可以直接在release v1.1上修改

把这个注释拉开
https://github.com/Yutong-gannis/ETSAuto/blob/c1465fc90ffef8020d079522c81e02e5c609c251/script/main.py#L151-L156

这行代码注释掉
https://github.com/Yutong-gannis/ETSAuto/blob/c1465fc90ffef8020d079522c81e02e5c609c251/script/main.py#L181-L182

只是之前有朋友反应车道线权重有问题，就把车道线相关的功能关掉了，你先试试

**Reply by Lyric0620** [reply]
branch的v1.1也是可以的，如果车道线检测不能用，你只能参考这个仓库自己转pt权重了
https://github.com/xuanandsix/CLRNet-onnxruntime-and-tensorrt-demo

**Reply by Shzzzz** [reply]
![image](https://github.com/Yutong-gannis/ETSAuto/assets/111844969/a1f0bbf5-094b-48ff-9a87-4c38d9683e5c)
跑了一小会，就出问题终止程序了

**Reply by Lyric0620** [reply]
对应函数照这样修改
https://github.com/Yutong-gannis/ETSAuto/blob/2883ed30b63d9a8d5b58ab2a7e0b0ffcf369df81/Planning/FSMPlaning.py#L39-L82

**Comment by Shzzzz** [comment]
刚刚测试了一下2.2.1版本，好像车辆自动驾驶的时候，如果导航路线有直角的转弯的地方，转不过去，他会走直线。

**Reply by Lyric0620** [reply]
目前版本只提供了车道保持和辅助变道控制。因为之前版本的导航线路的获取尺度有问题，所以还没有用到导航信息

**Comment by eiyooooo** [comment]
有没有考虑尝试一下用YOLOP呢😁

**Reply by Lyric0620** [reply]
我尝试过HybridNets，原论文的单类别检测效果可以，但是多类别检测时效果并不理想，并且车道线分割的范围也比较有限

**Comment by IDisNotDefined** [comment]
大佬，麻烦问一下我跑你的2.2.1版本，为什么车辆可以自动往前走，但是识别不出来其他车辆和道路？

**Reply by IDisNotDefined** [reply]
现在能检测出车了，但是检测不到道路

**Reply by IDisNotDefined** [reply]
大佬方便提供一下2.2.1版本里requirements里各个包的版本吗，我怀疑是我版本有冲突

**Reply by Lyric0620** [reply]
@IDisNotDefined 有报错吗

**Reply by IDisNotDefined** [reply]
没有收到报错

**Reply by Lyric0620** [reply]
检查一下五个程序是否都打开了

**Reply by IDisNotDefined** [reply]
运行的时候五个页面都有打开，单独开也没有报错。理论上我车能检测到的话代码应该就没什么问题，但车道就是检测不到，不知道为什么。排查也没看出什么问题，感觉可能还是环境里的包有冲突或者等级不对。大佬方便给下你2.2.1版本的requirement里各个包的版本号吗

**Reply by Lyric0620** [reply]
包冲突一般会报错的，我的numpy是1.24，其它的包都不太看版本。目前可以正常控制吗。
你可以先把这里的line_l打印一下，看看是否有值
https://github.com/Yutong-gannis/ETSAuto/blob/99a4a5ae9dfa988eb6aee2cf4c0c98dfd90217d4/Selfdrive/Perception/LaneDetection/lanedetector.py#L216

**Reply by IDisNotDefined** [reply]
大佬还想问一下，sdk是必须放在win_x86下吗，我放在x86下打开游戏不报警告，放在x64下才有警告。想问下sdk的放置位置是会影响读取数据吗

**Reply by Lyric0620** [reply]
你可以看一下log里的condition数据是否正常

**Reply by IDisNotDefined** [reply]
大佬请问log在哪看呀 我用的2.2.1版本 没找到日志。以及我运行的时候打开的五个窗口有的有报错
EOFError: D:\\ETSAuto-2.2.1\Selfdrive\Message/temp/option.pkl
lost trajectory data
lost trajectory data
lost trajectory data
lost trajectory data
lost trajectory data
lost trajectory data  请问这可能是什么原因造成的啊

**Reply by Lyric0620** [reply]
log应该保存在log文件夹下，这种情况可能是没检测出车道线，也可能perception没有在工作

**Comment by IDisNotDefined** [comment]
大佬我换了2.2.2的版本去运行，但速度检测一直是0，查日志没有报错但检测到的速度也是0，该怎么解决呀
![image](https://github.com/Yutong-gannis/ETSAuto/assets/141537568/15561f7c-f15b-4407-b87c-3344105a2788)

2024-02-05 12:19:32.678 | ConditionData | hardware_condition:update:68 - Hardware Condition: 
cpu_freq_current: 1000.0	 cpu_usage: 0.075
mem_free: 98.94608306884766	 mem_used: 28.713882446289062	 mem_percentage: 0.225
gpu_load: 0.39	 gpu_mem_free: 8530.0	 gpu_mem_used: 2510.0	 gpu_temp: 36.0
2024-02-05 12:19:32.681 | ConditionInfo | hardware_condition:update:75 - Condition update finish.
2024-02-05 12:19:32.683 | ConditionInfo | hardware_condition:publish:87 - Hardware condition publish finish.
2024-02-05 12:19:32.684 | ConditionInfo | __main__:run:47 - Loop time: 0.09099268913269043
2024-02-05 12:19:32.685 | ConditionData | truck_condition:update:44 - Truck condition:
 x: 0	 y: 0	 theta: 0.0	 dx: 0.0	 dy: 0.0	 dtheta: 0.0	 ang: 0	 acc: [0.0, 0.0]
 wheelbase: 8.0	 ld: 20	 lf: 3.63	 speed: 0.0	 speedlimit: 0.0	 overspeed:False	 dt:0.04
2024-02-05 12:19:32.686 | ConditionInfo | truck_condition:update:49 - Condition update finish.
2024-02-05 12:19:32.689 | ConditionInfo | truck_condition:publish:70 - Truck condition publish finish.
